osu.Game/Screens/BackgroundScreenStack.cs

```diff
@@ -33,5 +34,7 @@ public bool Push(BackgroundScreen? screen)
             base.Push(screen);
             return true;
         }
+
+        internal void ScheduleToTransitionEnd(Action action) => Scheduler.AddDelayed(action, BackgroundScreen.TRANSITION_LENGTH);
```

Comment:
Why does this exist?

Reply:
Right, I forgot to comment on this.

This all starts from wanting to have the storyboard stay present for as long as the storyboard background screen is visible during the suspension transition. The moment the screen is completely gone from the user's eyes, it'll also no longer be present in the hierarchy (talking about `BackgroundScreenDefault`).

Therefore, the only method I see possible to remove the storyboard after the transition finishes is to schedule the removal at a higher level that persists across transitions, and the closest one would be the `BackgroundScreenStack`.

I've touched on this briefly in [discord](https://discord.com/channels/188630481301012481/589331078574112768/1172171519338749952), as I was hoping I could schedule locally inside `BeatmapBackgroundWithStoryboard` and have it execute within the frame that the background screen is dead before it is marked so.

Reply:
I've added detailed explanation for this (see https://github.com/ppy/osu/pull/25400/commits/96da7a07bbd79bc67cbc58196bb38cd6c835c83a), give it a read and see what you think. I might be reinventing what `DelayedLoadUnloadWrapper` tries to do, but it doesn't look to work upon a simple try. Here's the diff I tried:
```diff
diff --git a/osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs b/osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs
index 784c8e4b44..6342b5669c 100644
--- a/osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs
+++ b/osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs
@@ -2,7 +2,6 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System.Collections.Generic;
-using System.Diagnostics;
 using System.Threading;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
@@ -21,7 +20,7 @@ public partial class BeatmapBackgroundWithStoryboard : BeatmapBackground
     {
         private readonly InterpolatingFramedClock storyboardClock;
 
-        private AudioContainer storyboardContainer = null!;
+        // private AudioContainer storyboardContainer = null!;
         private DrawableStoryboard? drawableStoryboard;
         private CancellationTokenSource? loadCancellationSource = new CancellationTokenSource();
 
@@ -31,7 +30,7 @@ public partial class BeatmapBackgroundWithStoryboard : BeatmapBackground
         [Resolved]
         private IBindable<IReadOnlyList<Mod>> mods { get; set; } = null!;
 
-        public BeatmapBackgroundWithStoryboard(WorkingBeatmap beatmap, string fallbackTextureName = "Backgrounds/bg1")
+        public BeatmapBackgroundWithStoryboard(WorkingBeatmap beatmap, string fallbackTextureName = @"Backgrounds/bg1")
             : base(beatmap, fallbackTextureName)
         {
             storyboardClock = new InterpolatingFramedClock();
@@ -40,59 +39,86 @@ public BeatmapBackgroundWithStoryboard(WorkingBeatmap beatmap, string fallbackTe
         [BackgroundDependencyLoader]
         private void load()
         {
-            AddInternal(storyboardContainer = new AudioContainer
-            {
-                RelativeSizeAxes = Axes.Both,
-                Volume = { Value = 0 },
-            });
-
-            LoadStoryboard(false);
-        }
-
-        public void LoadStoryboard(bool async = true)
-        {
-            Debug.Assert(drawableStoryboard == null);
+            // AddInternal(storyboardContainer = new AudioContainer
+            // {
+            //     RelativeSizeAxes = Axes.Both,
+            //     Volume = { Value = 0 },
+            // });
 
             if (!Beatmap.Storyboard.HasDrawable)
                 return;
 
-            drawableStoryboard = new DrawableStoryboard(Beatmap.Storyboard, mods.Value)
-            {
-                Clock = storyboardClock
-            };
+            DelayedLoadUnloadWrapper wrapper;
 
-            if (async)
-                LoadComponentAsync(drawableStoryboard, finishLoad, (loadCancellationSource = new CancellationTokenSource()).Token);
-            else
+            AddInternal(wrapper = new DelayedLoadUnloadWrapper(() => new AudioContainer
+            {
+                RelativeSizeAxes = Axes.Both,
+                Volume = { Value = 0 },
+                Alpha = 0f,
+                Child = drawableStoryboard = new DrawableStoryboard(Beatmap.Storyboard, mods.Value)
+                {
+                    Clock = storyboardClock,
+                }
+            }, 0, 0)
             {
-                LoadComponent(drawableStoryboard);
-                finishLoad(drawableStoryboard);
-            }
+                RelativeSizeAxes = Axes.Both,
+            });
 
-            void finishLoad(DrawableStoryboard s)
+            wrapper.DelayedLoadComplete += d =>
             {
                 if (Beatmap.Storyboard.ReplacesBackground)
                     Sprite.FadeOut(BackgroundScreen.TRANSITION_LENGTH, Easing.InQuint);
 
-                storyboardContainer.FadeInFromZero(BackgroundScreen.TRANSITION_LENGTH, Easing.OutQuint);
-                storyboardContainer.Add(s);
-            }
+                d.FadeIn(BackgroundScreen.TRANSITION_LENGTH, Easing.OutQuint);
+            };
+
+            // LoadStoryboard(false);
         }
 
-        public void UnloadStoryboard()
+        public void LoadStoryboard(bool async = true)
         {
-            if (drawableStoryboard == null)
-                return;
-
-            loadCancellationSource?.Cancel();
-            loadCancellationSource = null;
-
-            // clear is intentionally used here for the storyboard to be disposed asynchronously.
-            storyboardContainer.Clear();
-
-            drawableStoryboard = null;
+            // Debug.Assert(drawableStoryboard == null);
+            //
+            // if (!Beatmap.Storyboard.HasDrawable)
+            //     return;
+            //
+            // drawableStoryboard = new DrawableStoryboard(Beatmap.Storyboard, mods.Value)
+            // {
+            //     Clock = storyboardClock
+            // };
+            //
+            // if (async)
+            //     LoadComponentAsync(drawableStoryboard, finishLoad, (loadCancellationSource = new CancellationTokenSource()).Token);
+            // else
+            // {
+            //     LoadComponent(drawableStoryboard);
+            //     finishLoad(drawableStoryboard);
+            // }
+            //
+            // void finishLoad(DrawableStoryboard s)
+            // {
+            //     if (Beatmap.Storyboard.ReplacesBackground)
+            //         Sprite.FadeOut(BackgroundScreen.TRANSITION_LENGTH, Easing.InQuint);
+            //
+            //     storyboardContainer.FadeInFromZero(BackgroundScreen.TRANSITION_LENGTH, Easing.OutQuint);
+            //     storyboardContainer.Add(s);
+            // }
+        }
 
-            Sprite.Alpha = 1f;
+        public void UnloadStoryboard()
+        {
+            // if (drawableStoryboard == null)
+            //     return;
+            //
+            // loadCancellationSource?.Cancel();
+            // loadCancellationSource = null;
+            //
+            // // clear is intentionally used here for the storyboard to be disposed asynchronously.
+            // storyboardContainer.Clear();
+            //
+            // drawableStoryboard = null;
+            //
+            // Sprite.Alpha = 1f;
         }
 
         protected override void LoadComplete()

```

Reply:
I honestly I think I'd rather have the background fade itself out a little faster in order so that it can clean itself up while it's present than have this. It's just so ugly I can't overlook it.

Not sure if everyone will agree with me. @peppy would be interested as to whether you do.

Reply:
In fact, to offer something material, I'd much rather have something like this than what is in the PR currently:

```diff
diff --git a/osu.Game/Screens/BackgroundScreenStack.cs b/osu.Game/Screens/BackgroundScreenStack.cs
index 562b212561..99ca383b9f 100644
--- a/osu.Game/Screens/BackgroundScreenStack.cs
+++ b/osu.Game/Screens/BackgroundScreenStack.cs
@@ -1,12 +1,9 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-using System;
 using System.Collections.Generic;
 using osu.Framework.Graphics;
 using osu.Framework.Screens;
-using osu.Framework.Threading;
-using osu.Game.Screens.Backgrounds;
 
 namespace osu.Game.Screens
 {
@@ -36,20 +33,5 @@ public bool Push(BackgroundScreen? screen)
             base.Push(screen);
             return true;
         }
-
-        /// <summary>
-        /// Schedules a delegate to run after 500ms, the time length of a background screen transition.
-        /// This is used in <see cref="BackgroundScreenDefault"/> to dispose of the storyboard once the background screen is completely off-screen.
-        /// </summary>
-        /// <remarks>
-        /// Late storyboard disposals cannot be achieved with any local scheduler from <see cref="BackgroundScreenDefault"/> or any component inside it,
-        /// due to the screen becoming dead at the moment the transition finishes. And, on the frame that it is dead on, it will not receive an <see cref="Drawable.UpdateSubTree"/>,
-        /// therefore not guaranteeing to dispose the storyboard at any period of time close to the end of the transition.
-        /// This might require reconsideration framework-side, possibly exposing a "death" event in <see cref="Screen"/> or all <see cref="Drawable"/>s in general.
-        /// </remarks>
-        /// <param name="action">The delegate </param>
-        /// <returns></returns>
-        /// <seealso cref="BackgroundScreen.TRANSITION_LENGTH"/>
-        internal ScheduledDelegate ScheduleUntilTransitionEnd(Action action) => Scheduler.AddDelayed(action, BackgroundScreen.TRANSITION_LENGTH);
     }
 }
diff --git a/osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs b/osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs
index 4583b3e4d6..e46b92795a 100644
--- a/osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs
+++ b/osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs
@@ -9,6 +9,7 @@
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
 using osu.Framework.Logging;
+using osu.Framework.Platform;
 using osu.Framework.Screens;
 using osu.Framework.Threading;
 using osu.Framework.Utils;
@@ -36,6 +37,9 @@ public partial class BackgroundScreenDefault : BackgroundScreen
         [Resolved]
         private IBindable<WorkingBeatmap> beatmap { get; set; }
 
+        [Resolved]
+        private GameHost gameHost { get; set; }
+
         protected virtual bool AllowStoryboardBackground => true;
 
         public BackgroundScreenDefault(bool animateOnEnter = true)
@@ -81,7 +85,7 @@ public override void OnSuspending(ScreenTransitionEvent e)
             Debug.Assert(backgroundScreenStack != null);
 
             if (background is BeatmapBackgroundWithStoryboard storyboardBackground)
-                storyboardUnloadDelegate = backgroundScreenStack.ScheduleUntilTransitionEnd(storyboardBackground.UnloadStoryboard);
+                storyboardUnloadDelegate = gameHost.UpdateThread.Scheduler.AddDelayed(storyboardBackground.UnloadStoryboard, TRANSITION_LENGTH);
 
             base.OnSuspending(e);
         }

```

because contrary to what this PR does right now, that at least doesn't produce any new weird hooks out of thin air.

As for DLUW, that probably doesn't work because DLUW relies on the thing being unloaded moving *physically* off screen.

Reply:
I definitely prefer this proposal.

Reply:
I've pushed the aforementioned patch.